python/tskit/trees.py

```diff
@@ -2863,9 +2873,27 @@ def haplotypes(self, impute_missing_data=False):
         :raises: LibraryError if missing data is present and impute_missing_data
             is False
         """
-        hapgen = _tskit.HaplotypeGenerator(self._ll_tree_sequence, impute_missing_data)
-        for j in range(self.num_samples):
-            yield hapgen.get_haplotype(j)
+        H = np.empty((self.num_samples, self.num_sites), dtype=np.int8)
+        for var in self.variants(impute_missing_data=impute_missing_data):
+            # allow an extra allele at the end for missing data == -1
+            # filling with -1 values will ensure an error is raised if unassigned alleles
+            # are referenced and converted back to ascii
+            alleles = np.full((len(var.alleles) + 1), NULL, dtype=np.int8)
```

Comment:
I don't think we need the ``+ 1`` here: doesn't the ``var.alleles`` list end in ``None`` if there's missing data present?

Also, can't we just detect the presence of missing data here from the variant object, rather than depending on the Unicode error below?

Reply:
Yeah, we can use the [has_missing_data](https://tskit.readthedocs.io/en/latest/python-api.html#tskit.Variant) attribute.

Reply:
Yes, I think it *should* - I was probably being paranoid that it might not in all cases. But you are right, simpler just to cut this, and assume that the tests will catch any mistaken assumptions.

Re the unicode error, we would have to check that there were no `-1` values in the genotypes array, which is potentially quite a large array to check through, so I thought it would be better done when we have to convert it anyway. Alternatively, we could see if there was a `None` in the list of `var.alleles`, but I don't know if that is still present when e.g. `impute_missing_data=True`.

Reply:
> Yeah, we can use the [has_missing_data](https://tskit.readthedocs.io/en/latest/python-api.html#tskit.Variant) attribute.

Ah, didn't know about that. Perfect. Will recode it.